entity.go

```diff
@@ -101,15 +101,29 @@ func (lr *limitedReader) Read(p []byte) (int, error) {
 	return n, err
 }
 
-// Read reads a message from r. The message's encoding and charset are
-// automatically decoded to raw UTF-8. Note that this function only reads the
-// message header.
-//
-// If the message uses an unknown transfer encoding or charset, Read returns an
-// error that verifies IsUnknownCharset or IsUnknownEncoding, but also returns
-// an Entity that can be read.
-func Read(r io.Reader) (*Entity, error) {
-	lr := &limitedReader{R: r, N: maxHeaderBytes}
+type ReadOptions struct {
+	// MaxHeaderBytes limits the maximum permissible size of a message header
+	// block. If exceeded, an error will be returned.
+	// -1 == no limit, 0 == default value (1MB)
+	MaxHeaderBytes int64
+}
+
+// checkValues will modify the ReadOptions struct setting defaults or special
+// values. Must be called before the ReadOptions struct is used.
+func (o *ReadOptions) checkValues() {
+	if o.MaxHeaderBytes == 0 {
+		o.MaxHeaderBytes = defaultMaxHeaderBytes
+	} else if o.MaxHeaderBytes == -1 {
+		o.MaxHeaderBytes = math.MaxInt64
+	}
+}
+
+// ReadWithOptions see Read, but allows overriding some parameters with
+// ReadOptions.
+func ReadWithOptions(r io.Reader, opts ReadOptions) (*Entity, error) {
```

Comment:
In general options are a pointer to a struct (see other examples in the stdlib).

Reply:
This was actually deliberate - because the (currently named) `checkValues()` func mutates the struct, I didn't want to do that to a struct that I didn't own, it seems unexpected. By taking a value here, we get a copy that we are free to mutate.

Reply:
I've changed this to a pointer receiver in the latest commit, resolving this convo.